### PR TITLE
golint: Fix issues in pkg/nat

### PR DESCRIPTION
**Pull request by dave-tucker**
Updates #14756

Signed-off-by: Dave Tucker dt@docker.com

### CODE REVIEW

**Comment by dave-tucker** [comment]
Ok so I have a feeling that the issues in CI are related to the simply changing `HostIp` to `HostIP` in the `PortBinding` struct.
As we are using the names verbatim when marshalling to JSON, a simple case change will cause a whole world of pain :cry: 
I've added a tag to the struct to preserve the old-style JSON, but we should probably use something more JSONy like `host-ip` and/or add tags to all the external facing structs but that is way outside the scope of this patch...

File: pkg/nat/nat.go [hunk 1]
```diff
@@ -13,26 +13,33 @@ import (
 )
 
 const (
-	PortSpecTemplate       = "ip:hostPort:containerPort"
+	// PortSpecTemplate is the template for port specifications
```
Comment (stevvooe) [inline pkg/nat/nat.go]: Format? What do they mean? How should it be used?

File: pkg/nat/nat.go [hunk 2]
```diff
@@ -13,26 +13,33 @@ import (
...
+	PortSpecTemplate = "ip:hostPort:containerPort"
+	// PortSpecTemplateFormat is a list of valid formats for port specifications
 	PortSpecTemplateFormat = "ip:hostPort:containerPort | ip::containerPort | hostPort:containerPort | containerPort"
 )
 
+// PortBinding represents a binding between a Host IP address and a Host Port
 type PortBinding struct {
-	HostIp   string
+	HostIP   string `json:"HostIp"`
 	HostPort string
```
Comment (stevvooe) [inline pkg/nat/nat.go]: Document fields.

File: pkg/nat/nat.go [hunk 3]
```diff
@@ -13,26 +13,33 @@ import (
...
 }
 
+// PortMap is a collection of PortBinding indexed by Port
 type PortMap map[Port][]PortBinding
 
+// PortSet is a collection of structs indexed by Port
 type PortSet map[Port]struct{}
 
-// 80/tcp
+// Port is a string representaiton of a port, for example "80/tcp"
```
Comment (stevvooe) [inline pkg/nat/nat.go]: `representation`

This is really port and layer 4 protocol.

File: pkg/nat/sort.go [hunk 4]
```diff
@@ -26,6 +26,7 @@ func (s *portSorter) Less(i, j int) bool {
 	return s.by(ip, jp)
 }
 
+// Sort sorts a list of ports using the provided predicate
 func Sort(ports []Port, predicate func(i, j Port) bool) {
```
Comment (stevvooe) [inline pkg/nat/sort.go]: What invariant should be enforced by `predicate`? `sort.Sort` isn't necessarily stable, so `predicate` _must_ provide a total ordering over `ports` (I think, so doublecheck).

Unrelated to PR: this approach is not very idiomatic Go.

**Comment by dave-tucker** [comment]
Updated based on @stevvooe's review

**Comment by stevvooe** [comment]
@dave-tucker I would consider code that relies on untagged json structs for serialization to be broken. What tests are broken?

**Comment by dave-tucker** [comment]
@stevvooe failure was in the integration tests
for example, this one in `docker-py`... https://github.com/docker/docker-py/blob/8a87001d09852058f08a807ab6e8491d57ca1e88/tests/integration_test.py#L605

My hackaround feels wrong on many levels but I'm not really sure what to do here :confused:

**Comment by stevvooe** [comment]
@dave-tucker I see. This is very unfortunate that we let an internal struct escape into API. Your fix is acceptable.

LGTM

**Comment by jfrazelle** [comment]
LGTM

**Pull request merged**